packetbeat/protos/drda/drda.go

```diff
@@ -0,0 +1,572 @@
+package drda
+
+import (
+	"fmt"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+	"strings"
+	"time"
+
+	"github.com/Intermernet/ebcdic"
+	"github.com/elastic/beats/packetbeat/procs"
+	"github.com/elastic/beats/packetbeat/protos"
+	"github.com/elastic/beats/packetbeat/protos/tcp"
+	"github.com/elastic/beats/packetbeat/publish"
+	"github.com/urso/ucfg"
+)
+
+/*
+
+Parse DRDA Protocol used by IBM DB2, Informix or Apache Derby (and possibly other databases)
+
+Limitations:
+- Raw message not supported (guess that makes not much sense here)
+- No unit tests yet
+- Too few system tests
+- Check that all codepoints are covered (also "esoteric" like those from IMS)
+
+Dependencies:
+- https://github.com/Intermernet/ebcdic
+
+*/
+
+type parseState int
+
+const (
+	drdaStateDDM parseState = iota
+	drdaStateParameter
+)
+
+var stateStrings []string = []string{
+	"DDM",
+	"Parameter",
+}
+
+func drdaAbbrev(codepoint uint16) string {
+	abbrev := drda_abbrev[codepoint]
+
+	if abbrev == "" {
+		return fmt.Sprint("unknown_", codepoint)
+	}
+
+	return abbrev
+}
+
+func (state parseState) String() string {
+	return stateStrings[state]
+}
+
+func (drda *Drda) getTransaction(k common.HashableTcpTuple) *DrdaTransaction {
+	v := drda.transactions.Get(k)
+	if v != nil {
+		return v.(*DrdaTransaction)
+	}
+	return nil
+}
+
+func init() {
+	protos.Register("drda", New)
+}
+
+func New(
+	testMode bool,
+	results publish.Transactions,
+	cfg *ucfg.Config,
+) (protos.Plugin, error) {
+	p := &Drda{}
+	config := defaultConfig
+	if !testMode {
+		if err := cfg.Unpack(&config); err != nil {
+			return nil, err
+		}
+	}
+
+	if err := p.init(results, &config); err != nil {
+		return nil, err
+	}
+	return p, nil
+}
+
+func (drda *Drda) init(results publish.Transactions, config *drdaConfig) error {
+	drda.setFromConfig(config)
+
+	drda.transactions = common.NewCache(
+		drda.transactionTimeout,
+		protos.DefaultTransactionHashSize)
+	drda.transactions.StartJanitor(drda.transactionTimeout)
+	//drda.handleMysql = handleMysql
+	drda.results = results
+
+	return nil
+}
+
+func (drda *Drda) setFromConfig(config *drdaConfig) error {
+
+	drda.Ports = config.Ports
+	drda.Send_request = config.SendRequest
+	drda.Send_response = config.SendResponse
+	drda.transactionTimeout = time.Duration(config.TransactionTimeout) * time.Second
```

Comment:
after rebase: in master config.TransactionTimeout has become type 'time.Duration' which is natively supported by config.Unpack. update type + use simple assignment instead of multiply by seconds. Advantage is more flexibility in config file.